docs/csharp/nullable-attributes.md

```diff
@@ -0,0 +1,301 @@
+
+You can also specify that a return, or an `out` or `ref` argument isn't null even though the type is a nullable type. Consider a method that ensures an `Array` is large enough to hold a number of elements. If the input argument doesn't have capacity, the routine would allocate a new array and copy all the existing elements into it. If the input argument is `null`, the routine would allocate new storage. If there's sufficient capacity, the routine does nothing:
+
+```csharp
+public void EnsureCapacity<T>(ref T[] storage, int size)
+```
+
+You could call this routine as follows:
+
+```csharp
+// messages has the default value (null) when EnsureCapacity is called:
+EnsureCapacity<string>(ref messages, 10);
+// messages is not null.
+EnsureCapacity<string>(messages, 50);
+```
+
+After enabling null reference types, you want to ensure that the preceding code compiles without warnings. When the method returns, the `storage` argument is guaranteed to be not null. However, it's acceptable to call `EnsureCapacity` with a null reference. You can make `storage` a nullable reference type, and add the `NotNull` post-condition to the parameter declaration:
+
+```csharp
+public void EnsureCapacity<T>([NotNull]ref T[]? storage, int size)
+```
+
+The preceding code expresses the existing contract very clearly: Callers can pass a variable with the `null` value, but the return value is guaranteed to never be null. The `NotNull` attribute is most useful for `ref` and `out` arguments where `null` may be passed as an argument, but that argument is guaranteed to be not null when the method returns. You specify unconditional postconditions using the following attributes:
+
+- [MaybeNull](xref:System.Diagnostics.CodeAnalysis.MaybeNullAttribute): A non-nullable return value may be null.
+- [NotNull](xref:System.Diagnostics.CodeAnalysis.NotNullAttribute): A nullable return value will never be null.
+
+## Specify conditional post-conditions: `NotNullWhen` and `MaybeNullWhen`
+
+You're likely familiar with the `string` method <xref:System.String.IsNullOrEmpty(System.String)?DisplayProperty=nameWithType>. This method returns `true` when the argument isn't null, and not the empty string. It is a form of null-check: Callers don't need to null-check the argument if the method returns `false`. To make a method like this nullable aware, you'd set the argument to a nullable type, and add the `NotNullWhen` attribute:
+
+```csharp
+bool IsNullOrEmpty([NotNullWhen(false)]string? value);
+```
+
+That informs the compiler that any code where the return value is `false` need not be checked. The addition of the attribute informs the compiler's static analysis that `IsNullOrEmpty` performs the necessary null check: It returns `false` when the input argument is not `null`.
```

Comment:
In the last sentence the logic direction should be inverted. When input argument is an empty string (not `null`), it also returns `false`. However, if it returns 'false', we know that the input argument is not `null`:
```suggestion
That informs the compiler that any code where the return value is `false` need not be null-checked. The addition of the attribute informs the compiler's static analysis that `IsNullOrEmpty` performs the necessary null check: when it returns `false`, the input argument is not `null`.
```